File: frontend/src/types/expense-pool.ts
/**
 * ExpensePool domain types for expense categorization.
 *
 * These Zod schemas match exactly with backend/app/models/expense_pool.py.
 * Expense pools categorize GL entries and define gross-up behavior
 * for expense recovery calculations.
 */

import { z } from 'zod'

import { PoolTypeSchema } from './lease-recovery-profile'

/**
 * Helper for decimal string validation for gross-up targets.
 * Values must be between 0 and 1 (representing 0% to 100%).
 */
const grossUpTargetString = z
  .string()
  .regex(/^\d+(\.\d+)?$/, 'Must be a valid decimal number')
  .refine(
    (val) => {
      const num = parseFloat(val)
      return num >= 0 && num <= 1
    },
    { message: 'Gross-up target must be between 0 and 1' }
  )

/**
 * Full ExpensePool model from database.
 *
 * Contains all pool configuration including gross-up settings
 * used by the financial calculation engine.
 */
export const ExpensePoolSchema = z.object({
  id: z.string().uuid(),
  property_id: z.string().uuid(),
  name: z.string().min(1).max(100),
  pool_type: PoolTypeSchema,
  is_gross_up_applicable: z.boolean().default(true),
  gross_up_target: grossUpTargetString.nullable().optional(),
  description: z.string().max(500).nullable().optional(),
  parent_pool_id: z.string().uuid().nullable().optional(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
})

export type ExpensePool = z.infer<typeof ExpensePoolSchema>

/**
 * DTO for creating an expense pool.
 *
 * Requires property_id and pool configuration.
 */
export const ExpensePoolCreateSchema = z
  .object({
    property_id: z.string().uuid(),
    name: z.string().min(1).max(100),
    pool_type: PoolTypeSchema,
    is_gross_up_applicable: z.boolean().default(true),
    gross_up_target: grossUpTargetString.nullable().optional(),
    description: z.string().max(500).nullable().optional(),
    parent_pool_id: z.string().uuid().nullable().optional(),
  })
  .refine(
    (data) => {
      // gross_up_target should not be set when is_gross_up_applicable is false
      if (!data.is_gross_up_applicable && data.gross_up_target) {
        return false
      }
      return true
    },
    {
      message:
        'gross_up_target should not be set when is_gross_up_applicable is false',
      path: ['gross_up_target'],
    }
  )

export type ExpensePoolCreate = z.infer<typeof ExpensePoolCreateSchema>

/**
 * DTO for updating an expense pool.
 *
 * All fields are optional for partial updates.
 */
export const ExpensePoolUpdateSchema = z.object({
  name: z.string().min(1).max(100).optional(),
  pool_type: PoolTypeSchema.optional(),
  is_gross_up_applicable: z.boolean().optional(),
  gross_up_target: grossUpTargetString.nullable().optional(),
  description: z.string().max(500).nullable().optional(),
  parent_pool_id: z.string().uuid().nullable().optional(),
})

export type ExpensePoolUpdate = z.infer<typeof ExpensePoolUpdateSchema>

/**
 * Summary view of an expense pool for list displays.
 *
 * Includes essential fields and aggregated totals.
 */
export const ExpensePoolSummarySchema = z.object({
  id: z.string().uuid(),
  property_id: z.string().uuid(),
  name: z.string(),
  pool_type: PoolTypeSchema,
  is_gross_up_applicable: z.boolean(),
  total_amount: z
    .string()
    .regex(/^-?\d+(\.\d+)?$/, 'Must be a valid decimal number')
    .nullable()
    .optional(),
  entry_count: z.number().int().min(0).default(0),
})

export type ExpensePoolSummary = z.infer<typeof ExpensePoolSummarySchema>

/**
 * Expense pool with nested children for hierarchical display.
 *
 * Used for tree structure displays showing parent-child relationships.
 * Maximum depth: 2 levels (parent → child only).
 */
export const ExpensePoolWithChildrenSchema: z.ZodTypeAny = z.lazy(() =>
  z
    .object({
      id: z.string().uuid(),
      property_id: z.string().uuid(),
      name: z.string().min(1).max(100),
      pool_type: PoolTypeSchema,
      is_gross_up_applicable: z.boolean(),
      gross_up_target: grossUpTargetString.nullable().optional(),
      description: z.string().max(500).nullable().optional(),
      parent_pool_id: z.string().uuid().nullable().optional(),
      created_at: z.string().datetime(),
      updated_at: z.string().datetime(),
      children: z.array(ExpensePoolWithChildrenSchema).default([]),
      total_amount: z
        .string()
        .regex(/^-?\d+(\.\d+)?$/, 'Must be a valid decimal number')
        .nullable()
        .optional(),
    })
    .transform((data) => ({
      ...data,
      is_parent: data.children.length > 0,
      is_child:
        data.parent_pool_id !== null && data.parent_pool_id !== undefined,
    }))
)

export type ExpensePoolWithChildren = z.infer<
  typeof ExpensePoolWithChildrenSchema
>

/**
 * Helper to get display name for a pool type.
 */
export const getPoolTypeDisplayName = (poolType: string): string => {
  const displayNames: Record<string, string> = {
    operating: 'Operating Expenses',
    tax: 'Taxes',
    insurance: 'Insurance',
    capital: 'Capital Expenses',
    other: 'Other',
  }
  return displayNames[poolType] || poolType
}

File: backend/app/models/expense_pool.py
"""ExpensePool domain model for expense categorization.

The ExpensePool model configures expense buckets for each property,
determining how GL entries are categorized and whether gross-up
rules apply to the pool.
"""

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class ExpensePool(BaseModel):
    """Full expense pool model from database.

    Expense pools categorize GL entries and define gross-up behavior
    for expense recovery calculations.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    property_id: UUID = Field(description="Property this pool belongs to")
    name: str = Field(..., min_length=1, max_length=100, description="Pool name")
    pool_type: str = Field(
        ..., description="Category: operating, tax, insurance, capital, other"
    )
    is_gross_up_applicable: bool = Field(
        default=True, description="Whether gross-up applies to this pool"
    )
    gross_up_target: Decimal | None = Field(
        None,
        ge=Decimal("0"),
        le=Decimal("1"),
        description="Target occupancy for gross-up (e.g., 0.95)",
    )
    description: str | None = Field(None, max_length=500)
    parent_pool_id: UUID | None = Field(
        None, description="Parent pool for hierarchical grouping (max 2 levels)"
    )
    created_at: datetime
    updated_at: datetime

    @model_validator(mode="after")
    def validate_gross_up_target(self) -> "ExpensePool":
        """Validate gross_up_target is set when gross-up is applicable."""
        if self.is_gross_up_applicable and self.gross_up_target is None:
            # Allow None for now - target can be inherited from property default
            pass
        if not self.is_gross_up_applicable and self.gross_up_target is not None:
            # Clear gross_up_target if gross-up not applicable
            object.__setattr__(self, "gross_up_target", None)
        return self


class ExpensePoolCreate(BaseModel):
    """DTO for creating an expense pool.

    Used when setting up expense categorization for a property.
    """

    property_id: UUID
    name: str = Field(..., min_length=1, max_length=100)
    pool_type: str = Field(
        ..., description="Category: operating, tax, insurance, capital, other"
    )
    is_gross_up_applicable: bool = Field(default=True)
    gross_up_target: Decimal | None = Field(
        None,
        ge=Decimal("0"),
        le=Decimal("1"),
    )
    description: str | None = Field(None, max_length=500)
    parent_pool_id: UUID | None = Field(
        None, description="Parent pool ID for hierarchical grouping"
    )

    @model_validator(mode="after")
    def validate_gross_up_target(self) -> "ExpensePoolCreate":
        """Validate gross_up_target constraints."""
        if not self.is_gross_up_applicable and self.gross_up_target is not None:
            raise ValueError(
                "gross_up_target should not be set when is_gross_up_applicable is False"
            )
        return self


class ExpensePoolUpdate(BaseModel):
    """DTO for updating an expense pool.

    All fields are optional for partial updates.
    """

    name: str | None = Field(None, min_length=1, max_length=100)
    pool_type: str | None = None
    is_gross_up_applicable: bool | None = None
    gross_up_target: Decimal | None = Field(
        None,
        ge=Decimal("0"),
        le=Decimal("1"),
    )
    description: str | None = Field(None, max_length=500)
    parent_pool_id: UUID | None = Field(None, description="Parent pool ID")


class ExpensePoolSummary(BaseModel):
    """Summary view of an expense pool for list displays.

    Includes essential fields and aggregated totals.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    property_id: UUID
    name: str
    pool_type: str
    is_gross_up_applicable: bool
    total_amount: Decimal | None = Field(
        None, description="Sum of all GL entries in this pool"
    )
    entry_count: int = Field(default=0, ge=0, description="Number of GL entries")


class ExpensePoolWithChildren(BaseModel):
    """Expense pool with nested children for hierarchical display.

    Used for tree structure API responses showing parent-child relationships.
    Maximum depth: 2 levels (parent → child only).
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    property_id: UUID
    name: str = Field(..., min_length=1, max_length=100)
    pool_type: str
    is_gross_up_applicable: bool
    gross_up_target: Decimal | None = None
    description: str | None = None
    parent_pool_id: UUID | None = None
    created_at: datetime
    updated_at: datetime
    children: list["ExpensePoolWithChildren"] = Field(
        default_factory=list,
        description="Child pools (empty for leaf nodes)",
    )
    total_amount: Decimal | None = Field(
        None, description="Sum of pool expenses including children roll-up"
    )

    @property
    def is_parent(self) -> bool:
        """Check if this pool has children."""
        return len(self.children) > 0

    @property
    def is_child(self) -> bool:
        """Check if this pool has a parent."""
        return self.parent_pool_id is not None
